js/src/forum/fragments/AutocompleteDropdown.js

```diff
@@ -1,17 +1,16 @@
-import Component from 'flarum/Component';
+import Fragment from 'flarum/Fragment';
 
-export default class AutocompleteDropdown extends Component {
-  init() {
-    this.active = false;
-    this.index = 0;
-    this.keyWasJustPressed = false;
-  }
+export default class AutocompleteDropdown extends Fragment {
+  items = [];
+  active = false;
+  index = 0;
+  keyWasJustPressed = false;
 
   view() {
     return (
       <ul className="Dropdown-menu EmojiDropdown">
-        <li className="Dropdown-header" key="0">{app.translator.trans('flarum-emoji.forum.composer.type_to_search_text')}</li>
-        {this.props.items.map(item => <li key={item.attrs.key}>{item}</li>)}
+        <li className="Dropdown-header">{app.translator.trans('flarum-emoji.forum.composer.type_to_search_text')}</li>
+        {this.items.map(item => <li>{item}</li>)}
```

Comment:
Why were the `key` attributes removed? Because this now lives outside of Mithril's redraw cycle and there's no diffing?

Reply:
Mithril freaked out when we tried to set those. Is there a reason why we need them?

Reply:
They're important when diffing / (re-)rendering lists of things. See https://mithril.js.org/keys.html.

> Mithril freaked out

Can you add some more details, please? :pray:

Reply:
Unnecessary use of this attr seems to be discouraged: https://github.com/MithrilJS/mithril.js/issues/2510

While testing, I observed the following:

- Including `key="0"` results in the `TypeError: Vnodes must either always have keys or never have keys!` error
- Including the ones for the actual emojis does not, but there's no key attr showing up in the actual DOM.

What's the purpose of keeping this key attr? We don't seem to be using it anywhere...

Reply:
When you filter emojis, the images change place in the list. If an image was previously rendered at index 2, and is now at index 1, Mithril will update the `src` of the index 1 image and will load the image again.

With keys, Mithril will keep the image tags that are already loaded, improving rendering (no flickering for images previously loaded) and better performance (browser doesn't need to retrieve the image from cache again)

It's possible that `"0"` is not a valid Mithril key since it evaluates to falsy. We can use anything that can't be a key in the list. I think `"heading"` would work since all the other keys will contain a single emoji (I think).

Reply:
> but there's no key attr showing up in the actual DOM

That's normal. `key` is only used internally by Mithril and is not visible in the DOM.

Reply:
I don't think the issue is with it being falsy, but with the vnode itself ("heading" didn't work either). That being said, the concern about image srcs isn't really applicable here, right?

Reply:
Did you try your latest change? Now the `Dropdown-header` has no `key`, but the elements generated from te array do. So that should yield the same `TypeError` you mentioned above.

As @clarkwinkelmann, you should use a `key` for the heading that is not falsey and won't appear in the list of emojis. If I am not mistaken, the emojis from the list will all have a key containing colons, so Clark's suggestion of `"heading"` should do the trick.

Reply:
I did test, and heading didn't work.

Reply:
What does "didn't work" mean?

Reply:
Oh no, I'm one of those people.... 

Providing "heading", or any other value, results in the error I shared above.

Reply:
Ah, I think I understand now.

Our VDOM looks like this:

~~~
[
  {franz_suggested_key: 'heading'},
  [
    {key: ':grin:'},
    {key: ':gin:'},
  ],
]
~~~

The first `li` and the array of `li`s are not on the same level, as far as the VDOM is concerned. The nodes in the array all have keys, so that's why you're not seeing any error message anymore. Adding a key to the heading would require adding one to the *array object itself*.

I understand now. This is resolved. :smile: